enhancements/upgrade-readiness.md

```diff
@@ -0,0 +1,463 @@
+
+A core requirement of this enhancement is to provide operators with the ability to communicate to OLM that the operator should not be upgraded while a critical process is being executed. Consider the following scenario based on OLM today:
+
+> An operator installed with automatic updates is executing a critical process that could place the cluster in a bad state if interrupted. While this critical process is running, a new version of the operator becomes available. Since the operator was installed with automatic updates, OLM upgrades the operator without allowing the critical process to finish and places the cluster in an unrecoverable state.
+
+OLM can provide a significantly better user experience by creating a channel for the operator to communicate state that OLM cannot infer otherwise.
+
+### Goals
+
+- Provide operators with the means to explicitly communicate a condition that OLM cannot infer on its own.
+- Allow operators to define specific conditions which will influence OLM behavior, such as the "Upgrade Readiness" condition defined above.
+- Preserve existing behavior should the operator not communicate its upgradeable status.
+
+### Non-Goals
+
+- Provide operators with the means to cancel an upgrade that is in process.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As an operator author, I want my operator to be able to communicate conditions to OLM that it cannot infer otherwise.
+
+#### Story 2
+
+As an operator author, I want my operator to be able to communicate that OLM should not attempt to upgrade or uninstall my operator until I tell it otherwise.
+
+## Design Details
+
+Operators will communicate with OLM by setting [conditions](https://dev.to/maelvls/what-the-heck-are-kubernetes-conditions-for-4je7) on an owned CustomResource. Conditions are used by many Kubernetes Resources to provide users with additional insight into the state of the resource. For those unfamiliar with conditions, they typically follow the following format:
+
+| Field | Description |
+| ----- | ----------- |
+| Type |  A one-word description of the condition. |
+| Status | Status is the status of the condition. Can be True, False, Unknown. |
+| Reason | Unique, one-word, CamelCase reason for the condition's last transition. |
+| Message | Human-readable message indicating details about last transition. Typically used by cluster admins for debugging purposes. |
+| LastTransitionTime | Last time the condition transitioned from one status to another. |
+
+### OLM Supported Conditions
+
+Although an operator may set any condition on a CustomResource that it owns, there will be a subset of "OLM Supported Conditions" which will influence the behavior of OLM when present. The list of "OLM Supported Conditions" can be seen in the table below:
+
+| Condition Name | Effect |
+| -------------- | ------ |
+| NotInterruptible | When true, OLM will not upgrade the operator. The pod associated with the operator may still be killed by external forces.|
```

Comment:
Based on the description you provided, should this be `NotUpgradable` instead? As you noted, the the process can still be interrupted by external forces. 

Also, to minimize the potential for OLM to pickup conditions that are not intended for it by the operator (false positives) would it make sense to have an `OLM` prefix or something for the condition? Like `OLM.NotUpgradable = true`

Reply:
>Based on the description you provided, should this be NotUpgradable instead? As you noted, the the process can still be interrupted by external forces.

I had considered naming this condition `NotUpgradeable` but was concerned that we might want to use this condition for additional purposes down the road. I believed that `NotInterruptible` gave us enough wiggle room to allow operator to express their intent by setting the condition while giving OLM enough wiggle-room to expand on the behavior as we can make additional guarantees. An argument against this may be changing default upgrade behavior, but I assumed that the benefit of keeping "Operator Intent" clear was worth it.

>Also, to minimize the potential for OLM to pickup conditions that are not intended for it by the operator (false positives) would it make sense to have an OLM prefix or something for the condition? Like OLM.NotUpgradable = true

I believe this concern is discussed [here](https://github.com/operator-framework/enhancements/pull/23/files#diff-e72af68c33ea1f2583161d8a322dbf62R170). Basically, by requiring an operator to define the mapping there is no room for the conflict to occur.